spring-cloud-dataflow-scheduler-launcher/src/main/java/org/springframework/cloud/dataflow/scheduler/launcher/configuration/SchedulerTaskLauncher.java

```diff
@@ -0,0 +1,138 @@
+
+import org.springframework.cloud.dataflow.rest.client.DataFlowClientException;
+import org.springframework.cloud.dataflow.rest.client.TaskOperations;
+import org.springframework.cloud.dataflow.rest.resource.LauncherResource;
+import org.springframework.core.env.AbstractEnvironment;
+import org.springframework.core.env.EnumerablePropertySource;
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.MutablePropertySources;
+import org.springframework.hateoas.PagedModel;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Sends a request to the Spring Cloud Data Flow server to launch a task.
+ *
+ * @author Glenn Renfro
+ **/
+public class SchedulerTaskLauncher {
+
+	static final String TASK_PLATFORM_NAME = "spring.cloud.dataflow.task.platformName";
+
+	private static final Log log = LogFactory.getLog(SchedulerTaskLauncher.class);
+
+	private final String taskName;
+
+	private final String platformName;
+
+	private final SchedulerTaskLauncherProperties schedulerTaskLauncherProperties;
+
+	private final TaskOperations taskOperations;
+
+	private Environment environment;
+
+	public SchedulerTaskLauncher(String taskName, String platformName,
+			TaskOperations taskOperations, SchedulerTaskLauncherProperties schedulerTaskLauncherProperties,
+			Environment environment) {
+		Assert.hasText(taskName, "`taskName` cannot be null");
+		Assert.notNull(taskOperations, "`taskOperations` must not be null");
+		Assert.hasText(platformName, "`platformName` cannot be null");
+		Assert.notNull(schedulerTaskLauncherProperties, "`taskLauncherProperties` must not be null");
+		Assert.notNull(environment, "`environment` must not be null");
+		this.taskOperations = taskOperations;
+		this.taskName = taskName;
+		this.platformName = platformName;
+		this.schedulerTaskLauncherProperties = schedulerTaskLauncherProperties;
+		this.environment = environment;
+	}
+
+	public void launchTask(String... args) {
+		verifyTaskPlatform(this.taskOperations);
+		try {
+			log.info(String.format("Launching Task %s on the %s platform.", this.taskName, this.platformName));
+			this.taskOperations.launch(this.taskName, enrichDeploymentProperties(getDeploymentProperties()), Arrays.asList(args), null);
+		}
+		catch (DataFlowClientException e) {
+			throw new SchedulerTaskLauncherException(e);
+		}
+	}
+
+	private Map<String, String> enrichDeploymentProperties(Map<String, String> deploymentProperties) {
+		Map<String, String> enrichedProperties = new HashMap<>(deploymentProperties);
+		if (!deploymentProperties.containsKey(TASK_PLATFORM_NAME)) {
+			enrichedProperties.put(TASK_PLATFORM_NAME, this.platformName);
+		}
+		return enrichedProperties;
+	}
+
+
+	private Map<String, String> getDeploymentProperties() {
+		Map<String, String> props = new HashMap<>();
+		final String taskLauncherPrefix = this.schedulerTaskLauncherProperties.getTaskLauncherPropertyPrefix() + ".";
```

Comment:
I am not sure how this is supposed to work when we have the environment using the task deployment properties with `_` instead of `.`.

Reply:
In this case when multiple properties are sent to SchedulerTaskLauncher app it needs to know what properties need to be sent to SCDF for the task it is to launch. This prefix is used to mark those properties that are required for the app to be launched.   
For example if a user wants to schedule task definition `tstamp` (a timestamp app) with a property  of `app.timestamp.timestamp.format=YYYY` along with the other properties that are added by SCDF to be used by the SchedulerTaskLauncher.  We needed a method to mark this property as to be used for the `tstamp` definition launch.   So we append `tasklauncher.` to the property when it is scheduled with the platform's scheduler. i.e. it will look like `tasklauncher.app.timestamp.timestamp.format=YYYY`.   Thus when SchedulerTaskLauncher has been launched by the platform scheduler and it is associating the properties to be used for the SCDF API task launch call it will find these properties and marked with the `tasklauncher.` and associate them with the task launch.